presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java

```diff
@@ -282,11 +282,11 @@ private Type analyze(Expression expression, Scope scope, Context context)
     private class Visitor
             extends StackableAstVisitor<Type, Context>
     {
-        private final Scope scope;
+        private final Scope baseScope;
```

Comment:
This `Scope` is used to resolve `FieldReference` instances. Scope from context cannot be used for that.

Reply:
From semantic point of view `FieldReference` expressions should only be created for top level expressions (e.g: not in lambdas). Therefore context scope should be OK.

I guess the problem might be with `analyzeExpressionsWithInputs` which is used in `LocalExecutionPlanner` after `SymbolToInputRewriter`. However, I think that there should be another kind of expression (e.g: `ChannelReference`) for which type should be returned without using `Scope` at all (see https://github.com/prestodb/presto/issues/7398, `ChannelReference` would only be produced by SymbolToInputRewriter.

Reply:
I think you're right about where the problematic `FieldReference` instances come from. What you suggest sounds like a good direction for the future.

Reply:
At least minimally you should fix `ExpressionAnalyzer` with

> Alternatively, ExpressionAnalyzer might have optional Map<Integer, Type> inputMapping that would be used for FieldReference resolving (if present). The idea is that original (AST) FieldReferences would be translated to symbols so FieldReferences in LocalExecutionPlanner are only created by SymbolToInputRewriter. I think this is nicer temporary solution.

This will allow to void double scoping now in `ExpressionAnalyzer`.

Reply:
I added a comment explaining purpose of the `ExpressionAnalyzer.Visitor#baseScope` field.